tilings/algorithms/subclass_verification.py

```diff
@@ -0,0 +1,91 @@
+"""
+The main algorithm for detecting when the underlying ungridded permutations on a tiling
+are contained one of a given set of subclasses.
+
+NOTES:
+    - I suspect this algorithm will perform much better when <GriddedPermsOnTiling> is
+      replaced by a non-minimal version of <MinimalGriddedPerms>, both because of its
+      intense optimization and its breadth-first approach.
+    - This will be both more effective, and faster, if you don't do it until after any
+      points have been factored away.
+    - It may be possible to be smarter, and consider variable upper bounds on max_length
+      depending on which requirement in a list you're placing.
+"""
+
+from typing import TYPE_CHECKING, List, Set
+
+from permuta import Perm
+from tilings.algorithms import Factor, GriddedPermsOnTiling
+
+if TYPE_CHECKING:
+    from tilings import Tiling
+
+
+class SubclassVerificationAlgorithm:
+    def __init__(self, tiling: "Tiling", perms_to_check: Set[Perm]) -> None:
+        self.tiling = tiling
+        self.perms_to_check = perms_to_check
+
+    def quick_pare(self) -> Set[Perm]:
+
+        if len(self.tiling.requirements) == 0:
+            perms_leftover = set()
+
+            cell_basis = self.tiling.cell_basis()
+            all_bases = set(tuple(obs) for obs, _ in cell_basis.values())
+            max_base_len = {base: max(len(b) for b in base) for base in all_bases}
+            for perm in self.perms_to_check:
+                if not any(
+                    len(perm) < max_base_len[base] or perm.avoids(*base)
+                    for base in all_bases
+                ):
+                    perms_leftover.add(perm)
+            return perms_leftover
+
+        if (
+            len(self.tiling.requirements) == 1
+            and len(self.tiling.requirements[0]) == 1
+            and len(self.tiling.requirements[0][0]) == 1
+        ):
+            pos_cell = self.tiling.requirements[0][0].pos[0]
+            pos_basis = self.tiling.cell_basis()[pos_cell][0]
+            max_pos_basis_len = max(len(b) for b in pos_basis)
+            perms_leftover = set()
+            for perm in self.perms_to_check:
+                if not (len(perm) < max_pos_basis_len or perm.avoids(*pos_basis)):
+                    perms_leftover.add(perm)
+            return perms_leftover
+
+        return self.perms_to_check
+
+    def verified(self) -> bool:
+        if len(self.tiling.point_cells) > 0:
+            return False
+        if Factor(self.tiling).factorable():
+            return False
+
+        perms_to_check = self.quick_pare()
+        if len(perms_to_check) == 0:
+            return False
+
+        max_len_of_perms_to_check = max(map(len, perms_to_check))
+        max_length = (
+            self.tiling.maximum_length_of_minimum_gridded_perm()
+            + max_len_of_perms_to_check
```

Comment:
this can be -1, as you have to use at least one point of a minimum gridded perm, right?